picard/tagger.py

```diff
@@ -844,18 +838,45 @@ def autotag(self, objects):
     # =======================================================================
 
     def cluster(self, objs):
+        Thread(target=self._cluster, args=[objs]).start()
+
+    def _cluster(self, objs):
         """Group files with similar metadata to 'clusters'."""
         log.debug("Clustering %r", objs)
         if len(objs) <= 1 or self.unclustered_files in objs:
             files = list(self.unclustered_files.files)
         else:
             files = self.get_files_from_objects(objs)
-        for name, artist, files in Cluster.cluster(files, 1.0):
-            QtCore.QCoreApplication.processEvents()
+
+        # Temporarily disable left panel sorting to speed up related tasks
+        self.window.panel.set_sorting(False)
+
+        cluster_files = {}
+        for name, artist, files in Cluster.cluster(files, 1.0, self):
             cluster = self.load_cluster(name, artist)
             for file in sorted(files, key=attrgetter('discnumber', 'tracknumber', 'base_filename')):
-                file.move(cluster)
-                QtCore.QCoreApplication.processEvents()
+                if cluster not in cluster_files:
+                    cluster_files[cluster] = [file]
+                else:
+                    cluster_files[cluster].append(file)
+
+        pending_events = []
```

Comment:
We could use a `deque` instead of a list for this purpose: https://docs.python.org/3/library/collections.html#collections.deque

Reply:
OK